Initial commit for Scheduler object

milljm opened the pull request:
Initial commit of the scheduler object. We still initialize the runner with RunParallel.

Refs #9000

Commit:
Initial commit for Scheduler object

Refs #9000

permcody commented:
This is on the right track. I'm not going to merge this but keep going.

Commit:
use RunParallel as a plugin. Refs #9000

permcody commented:
Looking pretty good! You're getting this.

Commit:
Remove PBS

removes all references to PBS.
This makes way for adding PBS back in, as a plugin!

Refs #9000

Commit:
work in progress :)

Refs #9000

Commit:
Continued WIP

Refactored additional RunParallel code into Scheduler.
Added initial PBS commit. -sorry I didn't separate that

Refs #9000

milljm commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -72,8 +76,157 @@ def updateParams(self, tester):
+        json.dump(self.queue_data, self.queue_file, indent=2)
+
+        return output
+
```
This whole method should be in the base class... In fact I think all that belongs here is an re.complie of how to discover the returned job ID...

Commit:
Move checkRunnable into the Scheduler

Allow the Scheduler to perform checkRunnable. This allows
tests to be skipped if they were not previously launched
to a queue.

This commit also is the first commit with a working queue
submission to PBS

Refs #9000

Commit:
Handle tester.shouldExecute

When a tester does not want to execute the application
we need to instead return a command to execute so as to
allow the Queueing system to launch this test and retrieve
a Job ID

Refs #9000

Commit:
Fix moose/tests to work with Queueing

The following changes need to happen to make these tests work
with queueing options.

Refs #9000

Commit:
Allow tests to be skipped

Allow tests to be skipped if they can not support the Queueing
system.

Refs #9000

Commit:
Tests in modules that can not run with queueing

Modified these tests so they do not run when
using the Queue manager.

Refs #9000

Commit:
WIP to make moose/modules work with Queueing

The following changes are necessary in order to allow
queueing to function.

Refs #9000

Commit:
Launch all PBS jobs on HOLD

When using PBS options, place all launched jobs on hold as they
launch (qsub -h).

This fixes issues with pre-req tests completing before dependency
tests get launched. Because when this happens, the dependency test
will sit in the HOLD queue forever. Which will then require an
admin to release the job. I spoke to the torque guys, and this is
'operating as intended'.

Refs #9000

Commit:
Update json queue_file when checking status

Allow the json queue file to store current test
status information (maybe useful to someone out
there).

Pollish the comments a bit.

Refs #9000

moosebuild commented:
Job `PBS Test:linux-pbs` on 7c37d9c : invalidated by @brianmoose

JIT fix is in

Commit:
WIP for PBS queue

Work in progress commit trying to get the queueing
system up and running. This commit appears to function
correctly. Needs pollish.

Refs #9000

Commit:
Notify derived classes of status change

The QueueManager needed to know when the status of a test
was changed, and was not.

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on a1bc2d9 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Symlink causing duplicate tests

Record launched test information up at the os.walk level
to prevent duplicate testers from being created.

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on 543a1b2 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Allow queue_data to grow

Add the ability to create and update queue_data fields on the fly.
Add a few tester status methods to make life easier.
Clean up some stuff...
Added better warnings when things go wrong.

Refs #9000

milljm commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -54,17 +54,64 @@ def handleQueueLaunch(self, tester, output):
+                # There is current data we need to append to
+                if type(data[key]) == type([]) and type(value) == type([]):
+                    data_copy[key].extend(value)
+                    self.putData(job, **data_copy)
```
as we continue to modify data_copy, we can probably do putData all in one go after the for loop.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 5c98a95 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Add a reverse dependency resolver

When checking the status on tests, we need the ability
to determine 'what depends on me' function. We then
re-order the prereqs for these tests, and continue
to use the forward dependency resolver we have always
used.

Unit tests for this functionality will follow...

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on ad00559 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Move reverse dependency resolver to QueueManager

The reverse dependency resolver needs access to the
queue_data that is only handled by the QueueManager.

Due to the way the queue_data only knows about tests
that actually 'made it' to the QueueManager, we can
only modify prereqs on tests in the queue_data file.

I ran into an issue when modifying a tests prereqs to
include tests _not_ originally run (saved in the json
queue_data file) created cyclic dependency errors.

Better way? maybe.

Refs #9000

moosebuild commented:
Job `App tests:linux-gnu` on ad00559 : invalidated by @YaqiWang

Rattlesnake/MAMMOTH fixed

Commit:
Remove previous Reverse Dependency Resolver

Remove previous reverse dependency resolver in favor
of a new more betterer one.

Refs #9000

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -161,9 +161,6 @@ def findAndRunTests(self, find_only=False):
 
-                    # See if any tests have colliding outputs
-                    self.checkForRaceConditionOutputs(testers, dirpath)
-
```
we are already doing this when we launch tests

moosebuild commented:
Job `Moose Documentation:linux-clang` on 2865f83 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

permcody committed:
Address comments

refs #9164

Commit:
Handle queued skipped/pending jobs

Properly handle jobs in a queue that are being skipped or pending
when coming back around and checking their status. Jobs that were
being skipped due to dependencies found later on (during join())
were not properly being stored in the queue_data dictionary for
later use.

Refs #9000

Commit:
Allow failed dependents as 'skipped dependency'

revert to normal skipped dependency statuses for reverse
dependent failures.

Refs #9000

Commit:
Allow these tests to run with queueing options

The following tests were set not to run for reason due to dependency
issues while checking status. Now that has been solved. Allow these
tests to run, and if they still fail, fix the test...

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on de049b0 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Use previous status information

When checking status multiple times while using queueing features, use
the information provided by the json file to repeat the status of the
test.

While checking status during queueing operations, if a test is skipped
due to dependencies, verify that really is the case and alter the
status message if it is not. This was necessary when we (correctly)
began running the postProcess checks in reverse (reverse dependencies).

Refs #9000

milljm commented on modules/solid_mechanics/examples/bridge/bridge.e:
```diff
@@ -0,0 +1 @@
+../../../tensor_mechanics/examples/bridge/bridge.e
```
turns out we can not link the whole directory... Because that directory included a 'tests' file, which ended up getting copied to the new queue directory. Which created a duplicate test the next time the TestHarness was instructed to run.

milljm commented on modules/solid_mechanics/examples/bridge/fake_bridge.e:
```diff
@@ -0,0 +1 @@
+../../../tensor_mechanics/examples/bridge/fake_bridge.e
```
same comment above

milljm commented on modules/solid_mechanics/examples/bridge/tests:
```diff
@@ -5,18 +5,16 @@
-    cli_args = 'Mesh/file=bridge/fake_bridge.e'
+    cli_args = 'Mesh/file=fake_bridge.e'
     check_input = True
-    link_files = bridge
```
see above comments

milljm commented on modules/solid_mechanics/examples/bridge/tests:
```diff
@@ -5,18 +5,16 @@
-    cli_args = 'Mesh/file=bridge/fake_bridge.e'
+    cli_args = 'Mesh/file=fake_bridge.e'
     check_input = True
-    link_files = bridge
```
see above comments

Commit:
Accidentally left debugging code. Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on a284c7a wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Check for redirected output _correctly_

We were not doing this correctly. And not at all while using Queueing options.

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on 8e345e2 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Forgot to add methods to base class

Forgot to add the following two methods to the base class.

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on f41953b wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Not all tests can have redirected output

Remove the `if options.parallel` logic from redirected output. We can not
use parallel options alone to determine if a test should use redirected
output.

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on c168247 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Make module tests work with Queueing

Fix several tests which were writing (overwriting) to the same
output file as their prereq test. This was causing the prereq
test to fail the 'processResults' method.

Refs #9000

moosebuild commented:
Job `Moose Documentation:linux-clang` on fffcfe0 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Merge remote-tracking branch 'upstream/devel' into refactor-runparallel-9000

Commit:
Pollish the code a bit

Refactored the processResults method in the Queue plugins.
When checking status on previous tests, set the status bucket
directly, instead of logically.

Refs #9000

milljm commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -54,30 +55,36 @@ def handleQueueLaunch(self, tester, output):
+                     test_name      = tester.getTestName(),
+                     input_name     = self.options.input_file_name,
+                     status_message = tester.getStatusMessage(),
+                     status_bucket  = dict(tester.getStatus()._asdict()))
```
because I am now storing the actual status_bucket in the dictionary, we can consider removing all those individual test statuses from the dictionary (is_pending, did_pass, skipped, etc). However, because its a bit more expensive and code intensive to; create a bucket, and then compare that bucket to a tester bucket to figure out the same results as doing a:  `if dict['is_pending']:` I think we should leave this alone.

moosebuild commented:
Job `Moose Documentation:linux-clang` on a712406 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

Commit:
Use methods to gather tester statuses

Just a little bit of code cleanup to normalize the way we should be
asking the testers for some information.

Refs #9000

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -476,31 +472,31 @@ def handleTestResult(self, tester, output, result, start=0, end=0, add_to_table=
 
-        if self.options.verbose or (not did_pass and not self.options.quiet and not status == tester.bucket_pending):
+        if self.options.verbose or (self.options.quiet is not True and (tester.didFail() or tester.didDiff())):
             output = output.replace('\r', '\n')  # replace the carriage returns with newlines
```
_zero_ clue how that was working before.

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -476,31 +472,31 @@ def handleTestResult(self, tester, output, result, start=0, end=0, add_to_table=
-                self.writeFailedTest.write(tester.specs['test_name'] + '\n')
+        if not tester.isSkipped() and not tester.isPending():
+            if not tester.didPass() and not self.options.failed_tests:
+                self.writeFailedTest.write(tester.getTestName() + '\n')
```
:/

```python
if self.options.failed_tests is not True and (tester.didFail() or tester.didDiff()):
```

moosebuild commented:
Job `Moose Documentation:linux-clang` on 31fb5d0 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9001/site

milljm commented:
I am going to separate the commits into easier-to-manage PRs

#9238  and #9237
